### DISCUSSION TITLE
[Item] collect annotations by color(sort by color)

### DISCUSSION THREAD

**Discussion by windingwind** [windingwind/zotero-better-notes#227] [Note Templates]
## Details
Author: @windingwind
Discuss: #49 
Type: A

![175812654-0d2cf807-9967-4961-a4e8-00cad79893ab](https://user-images.githubusercontent.com/33902321/202897257-7a0e8d23-de13-4b7f-93b6-c1834d6e7d47.png)
![174488228-000257b4-10c4-47eb-a833-a6bd2516b5cf](https://user-images.githubusercontent.com/33902321/202897262-acc38409-ddaf-40f1-9406-ff5edc98dd5d.png)

## Template Code for Importing
```yaml
# This template is specifically for importing/sharing, using better 
# notes 'import from clipboard': copy the content and
# goto Zotero menu bar, click Edit->New Template from Clipboard.  
# Do not copy-paste this to better notes template editor directly.
name: "[item] collect annotations by color(sort by color) "
content: |-
  // @author windingwind
  // @link https://github.com/windingwind/zotero-better-notes/discussions/227
  ${{
    async function getAnnotationsByColor(item, colorFilter) {
      const annots = item.getAnnotations().filter(colorFilter);
  	return await Zotero.BetterNotes.api.convert.annotations2html(annots, { noteItem: targetNoteItem });
    }
  
    const attachments = Zotero.Items.get(topItem.getAttachments()).filter((i) =>
      i.isPDFAttachment() || i.isSnapshotAttachment() || i.isEPUBAttachment()
    );
    let res = "";
    const colors = ["#ffd400", "#ff6666", "#5fb236", "#2ea8e5", "#a28ae5", "#e56eee", "#f19837", "#aaaaaa"];
    const colorNames = ["Yellow", "Red", "Green", "Blue", "Purple", "Magenta", "Orange", "Gray"];
    for (let attachment of attachments) {
      res += `<h1>${attachment.getFilename()}</h1>`;
      for (let i in colors) {
        const renderedAnnotations = (
          await getAnnotationsByColor(
            attachment,
            (annot) => annot.annotationColor === colors[i]
          )
        );
        if (renderedAnnotations) {
          res += `<h2><p style="background-color:${colors[i]};">${colorNames[i]} Annotations</p></h2>\n${renderedAnnotations}`;
        }
      }
      const renderedAnnotations = (
        await getAnnotationsByColor(
          attachment,
          (annot) => !colors.includes(annot.annotationColor)
        )
      );
      if (renderedAnnotations) {
        res += `<h2><p>Other Annotations</p></h2>\n${renderedAnnotations}`;
      }
    }
    return res;
  }}$
```

**Comment by MengjieLi-Cora** [comment]
Hello, I am a beginner of Zotero. I want to collect annotations by color (sort by color). But code here seems not work in zotero 6. Same code, but colored annotations did not sorted. Is there any steps I have missed? I would appreciated if any related information you want to share with me.

**Reply by Superflash1** [reply]
I have the same problem.

**Comment by AlCATO42** [comment]
Hola Comparto el código de esta plantilla actualizado para que reconozca los nuevos colores que se agregaron al lector de Zotero.
${await new Promise(async (r) => {
  async function getAnnotation(item) {
    try {
      if (!item || !item.isAnnotation()) {
        return null;
      }
      let json = await Zotero.Annotations.toJSON(item);
      json.id = item.key;
      delete json.key;
      for (let key in json) {
        json[key] = json[key] || "";
      }
      json.tags = json.tags || [];
      return json;
    } catch (e) {
      Zotero.logError(e);
      return null;
    }
  }

  async function getAnnotationsByColor(_item, color) {
    const annots = _item
      .getAnnotations()
      .filter((_annot) => _annot.annotationColor === color);
    let annotations = [];
    for (let annot of annots) {
      const annotJson = await getAnnotation(annot);
      annotJson.attachmentItemID = _item.id;
      annotations.push(annotJson);
    }

    const editor =
      await Zotero.Knowledge4Zotero.knowledge.getWorkspaceEditorInstance();
    await editor.importImages(annotations);
    return Zotero.EditorInstanceUtilities.serializeAnnotations(annotations);
  }

  const attachments = Zotero.Items.get(topItem.getAttachments()).filter((i) =>
    i.isPDFAttachment()
  );
  let res = "";
  let colorYellow = "#ffd400";
  let colorRed = "#ff6666";
  let colorGreen = "#5fb236";
  let colorBlue = "#2ea8e5";
  let colorPerple = "#a28ae5";
  let colorMagenta = "#e56eee";
  let colorOrange = "#f19837";
  let colorGray = "#aaaaaa";
  for (let attachment of attachments) {
    res += `<h2><p style="background-color:${colorYellow};">Yellow Heading</p></h2>
    ${(await getAnnotationsByColor(attachment, colorYellow)).html}
    `;
    res += `<h2><p style="background-color:${colorRed };">Red Heading</p></h2>
    ${(await getAnnotationsByColor(attachment, colorRed )).html}
    `;
    res += `<h2><p style="background-color:${colorGreen };">Green Heading</p></h2>
    ${(await getAnnotationsByColor(attachment, colorGreen )).html}
    `;
    res += `<h2><p style="background-color:${colorBlue };">Blue Heading</p></h2>
    ${(await getAnnotationsByColor(attachment, colorBlue )).html}
    `;
    res += `<h2><p style="background-color:${colorPerple };">Perple Heading</p></h2>
    ${(await getAnnotationsByColor(attachment, colorPerple )).html}
    `;
    res += `<h2><p style="background-color:${colorMagenta };">Magenta Heading</p></h2>
    ${(await getAnnotationsByColor(attachment, colorMagenta )).html}
    `;
  res += `<h2><p style="background-color:${colorOrange };">Orange Heading</p></h2>
    ${(await getAnnotationsByColor(attachment, colorOrange )).html}
    `;
  res += `<h2><p style="background-color:${colorGray };">Gray Heading</p></h2>
    ${(await getAnnotationsByColor(attachment, colorGray )).html}
    `;
  }
  r(res);
})}

**Comment by Aries-Fu** [comment]
Hi, I met an issue that when I updated to the new version, I could not use this template. I tried to copy and paste this template again into the Template Editor, and name it with the prefix [item]. However, it still did not work. Can you tell me how to deal with this please? Thanks! @windingwind

**Comment by windingwind** [comment]
The template is updated for the latest beta version (1.1.4-18)

**Comment by zzlb0224** [comment]
这个功能分类整理太强了，感谢

**Comment by outsider15** [comment]
应用模板的时候显示错误：Template [item] collect annotations by color(sort by color)  Error: SyntaxError: missing : after property id，怎么解决，谢谢！

**Reply by windingwind** [reply]
更新插件。仅适用于最新beta版本（zotero 7）

**Reply by judywang1111** [reply]
更新了。还是不行

**Reply by windingwind** [reply]
@judywang1111 
1. 你需要描述怎么操作的，出了什么问题。看病不说病情，华佗也救不了
2. 是否按照文档描述的导入模板操作了
3. 当前插件和zotero什么版本

**Reply by Adammartain** [reply]
大佬好，我一时疏忽，没按文档描述操作，把collect annotations直接在模板编辑器里粘贴了，之后不论通过注释还是模板添加条目笔记时，都出现如下字段
![image](https://github.com/windingwind/zotero-better-notes/assets/56383761/053cf4fe-d0b9-49a5-8201-9e1bcd0ff0db)
求大佬帮忙看看na哪里出问题了呜呜呜，以后再也不不看文档就乱操作了

**Reply by windingwind** [reply]
删除这个模板，然后阅读一下模板开头几行

**Comment by yuancycle** [comment]
您好，模板报错SyntaxError: missing : after property id ，复制您上面的代码从剪切板导入新模板，zotero7 bn版本1.0.4，使用了几种导入模板到文件的方式都报错

**Reply by windingwind** [reply]
善用搜索，升级插件

**Comment by pinetreemoonlight** [comment]
Would it be possible to edit this template to 

1. show the annotations as bullet points under each heading
2. to remove the parentheses for each annotation (" ")
3. to remove the part in brackets for each annotation (author name, year and page number)? 
4. 
I have no idea how to do it myself, so I am asking kindly!

**Reply by windingwind** [reply]
you can ask for tools like GPT for help. sorry I do not have time to satisfy these customization requests.

**Reply by pinetreemoonlight** [reply]
I couldn't get it to work even with GPT. If anyone else can bother to try, I would appreciate it!

**Reply by pinetreemoonlight** [reply]
I figured it out. You have to go to Zotero Settings --> Advanced --> Config Editor and find `annotations.notesTemplates.highlight` and enter 
```html
<ul><li>{{highlight}}{{comment}}</li></ul>
```

**Comment by alic234** [comment]
大佬，能不能求一版按颜色获取 划了下划线 的笔记（高亮看的眼花）非常感谢！！！

**Reply by windingwind** [reply]
下划线应该也能用的？

**Reply by alic234** [reply]
我试了几篇都不行欸，只能获取高亮的文字，划了下划线得获取不了

**Comment by sBaydin** [comment]
Hello,
I am using >Zotero 6 and tried many thing about sorting with color but cant find any proper solition. Maybe some of guys sutruggling with the same thing as i did. So thats my solition:
1- On the setting find extensions.zotero.annotations.noteTemplates.highlight
2- Than paste this:

```
<!-- Sarı : #ffd400 renkli sarı, önemli notlar, kilit bilgiler veya öne çıkan detaylar için kullanabilirsin. -->
{{if color == '#ffd400'}}
<h4 style="color: #F8DE22; background-color: #EEF5FF;font-weight: bold;"><strong>🌟ℹ️ Important Note or Information:</strong></h4>
{{highlight}} {{citation}}
	{{if comment}}
	<blockquote>
		<p style="background-color: #FFF200; font-style: italic;">{{comment}}
	</blockquote>
	{{endif}}
	<hr>
<!-- Kırmızı : #ff6666 renkli kırmızı, özellikle dikkat edilmesi gereken noktaları vurgulamak için uygun olabilir. -->
{{elseif color == '#ff6666'}}
<h4 style="color: red; background-color: #EEF5FF;font-weight: bold;"><strong>⚠️❗ Point of Attention:</strong></h4>
{{highlight}} {{citation}}
	{{if comment}}
	<blockquote>
		<p style="background-color: #FFB3B3;font-style: italic;">{{comment}}
	</blockquote>
	{{endif}}
	<hr>
<!-- Yeşil : #5fb236 renkli yeşili, çözümler, olumlu bilgiler veya başarı öyküleri gibi pozitif içerikleri vurgulamak için kullanabilirsin. -->
{{elseif color == '#5fb236'}}
<h4 style="color: green; background-color: #EEF5FF;font-weight: bold;"><strong>💡✅ Solution or Positive Information:</strong></h4>
{{highlight}} {{citation}}
	{{if comment}}
	<blockquote>
		<p style="background-color: #AFD99B;font-style: italic;">{{comment}}
	</blockquote>
	{{endif}}
	<hr>
<!-- Mavi : #2ea8e5 renkli mavi, referanslar veya bağlantılar gibi ek bilgilere işaret etmek için kullanılabilir.-->
{{elseif color == '#2ea8e5'}}
<h4 style="color: blue; background-color: #EEF5FF;font-weight: bold;"><strong>📘🔗 References or Links:</strong></h4>
{{highlight}} {{citation}}
	{{if comment}}
	<blockquote>
		<p style="background-color: #97D4F2;font-style: italic;">{{comment}}
	</blockquote>
	{{endif}}
	<hr>
<!-- Mor : #a28ae5 renkli moru, makalenin ana konu başlıklarını veya ana fikrini vurgulamak için kullanabilirsin. -->
{{elseif color == '#a28ae5'}}
<h4 style="color: purple; background-color: #EEF5FF;font-weight: bold;"><strong>🔍📌 Main Topic or Main Idea:</strong></h4>
{{highlight}} {{citation}}
	{{if comment}}
	<blockquote>
		<p style="background-color: #D1C5F2;font-style: italic;">{{comment}}
	</blockquote>
	{{endif}}
	<hr>
<!-- Pembe : #e56eee renkli pembe, ilginç bilgileri, istatistikleri veya öne çıkan detayları vurgulamak için kullanabilirsin. -->
{{elseif color == '#e56eee'}}
<h4 style="color: magenta; background-color: #EEF5FF;font-weight: bold;"><strong>📈💖 Interesting Facts or Statistics:</strong></h4>
{{highlight}} {{citation}}
	{{if comment}}
	<blockquote>
		<p style="background-color: #F2B7F6;font-style: italic;">{{comment}}
	</blockquote>
	{{endif}}
	<hr>
<!-- Turuncu : #f19837 renkli turuncu, temel kavramlar, anahtar bilgiler veya önemli detaylar için kullanılabilir. -->
{{elseif color == '#f19837'}}
<h4 style="color: orange; background-color: #EEF5FF;font-weight: bold;"><strong>🔑📚 Key Information or Fundamental Concepts:</strong></h4>
{{highlight}} {{citation}}
	{{if comment}}
	<blockquote>
		<p style="background-color: #F8CC9B;font-style: italic;">{{comment}}
	</blockquote>
	{{endif}}
	<hr>
<!-- Gri : #aaaaaa renkli griyi, genel bilgileri, arka plan bilgilerini veya daha az vurgulanması gereken detayları belirtmek için kullanabilirsin. -->
{{elseif color == '#aaaaaa'}}
<h4 style="color: grey; background-color: #EEF5FF;font-weight: bold;"><strong>📖🔍 General Information or Background:</strong></h4>
{{highlight}} {{citation}}
	{{if comment}}
	<blockquote>
		<p style="background-color: #D5D5D5;font-style: italic;">{{comment}}
	</blockquote>
	{{endif}}
	<hr>
{{endif}}
```

That is Turkish writing. You can use a transaltor if you want and change it as you wish.

Meanwhile for expert users; if possible can you arrange it for sorting with colors.
Good luck

**Reply by windingwind** [reply]
I'm not sure if you want to share your workflow (which should be in a separate discussion) or if you are looking for help with Zotero's built-in note template (which should go to Zotero forums).

**Comment by sBaydin** [comment]
> I'm not sure if you want to share your workflow (which should be in a separate discussion) or if you are looking for help with Zotero's built-in note template (which should go to Zotero forums).

Sorry if i write it wrong place but actually i want help for this discussion.
I'm trying to group by colour in betternote but I couldn't do it. I finally read this method and arranged it myself as mentioned above. The reason I'm writing here is that I couldn't run the codes written as [item] in betternote. I can't understand if it's because I'm using Zotero 6 or if I'm making a mistake. The code I used for BetterNote is as follows:
```
<details>
    <summary>
        <h2 style="color: #004d99; background-color: #87cefa; margin: 0; padding: 8px">🌟 annotation</h2>
    </summary>
    <h3>🖊️ Writing</h3>
    <p></p>
    <h3>📈 Tables and Figures</h3>
    <p></p>
</details>
```
automatic retrieval of quotations as writing where it says [writing]
I want to have the sections that we have marked as fields where it says [tables and figures].

In addition to this, I am trying to arrange the subgroups of these groups according to the colours, i.e. yellow under one heading and red under one heading.
But I couldn't succeed somehow.
By the way, thanks for the quick response.

**Comment by huberdf** [comment]
论文过一段时间又看了一遍，又重新写了些highlighting和注释，笔记后面又更新了，笔记里的好像不能自动更新。

**Reply by fredericky123** [reply]
> 论文过一段时间又看了一遍，又重新写了些highlighting和注释，笔记后面又更新了，笔记里的好像不能自动更新。

期待能自动更新
